runtime/engine_context.h

```diff
@@ -0,0 +1,90 @@
+
+// Copyright 2013 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+#ifndef SHELL_COMMON_ENGINE_CONTEXT_H_
+#define SHELL_COMMON_ENGINE_CONTEXT_H_
+
+#include "flutter/common/task_runners.h"
+#include "flutter/fml/memory/weak_ptr.h"
+
+namespace flutter {
+
+class DartSnapshot;
+
+class SnapshotDelegate;
+
+class HintFreedDelegate;
+
+class IOManager;
+
+class SkiaUnrefQueue;
+
+class ImageDecoder;
+
+class ImageGeneratorRegistry;
+
+class VolatilePathTracker;
+
+/// @brief  The subset of engine-owned state which is passed through the
+///         runtime controller and accessible via the UIDartState.
+struct EngineContext {
+  EngineContext();
+
+  EngineContext(TaskRunners task_runners,
+                fml::WeakPtr<SnapshotDelegate> snapshot_delegate,
+                fml::WeakPtr<HintFreedDelegate> hint_freed_delegate,
+                fml::WeakPtr<IOManager> io_manager,
+                fml::RefPtr<SkiaUnrefQueue> unref_queue,
+                fml::WeakPtr<ImageDecoder> image_decoder,
+                fml::WeakPtr<ImageGeneratorRegistry> image_generator_registry,
+                std::string advisory_script_uri,
+                std::string advisory_script_entrypoint,
+                std::shared_ptr<VolatilePathTracker> volatile_path_tracker);
+
+  /// The task runners used by the shell hosting this runtime controller. This
+  /// may be used by the isolate to scheduled asynchronous texture uploads or
+  /// post tasks to the platform task runner.
+  TaskRunners task_runners;
```

Comment:
The problem is that in some places (like in the unittests) EngineContext is default constructed and then overwritten in somewhat arbitrary ways. Since there's a bunch of variation in what needs to be filled out based on the context of usage/what's being tested, I ended up dropping const in order to keep things consistent/avoid making special constructors to assign `task_runners`.

Reply:
Would it be simpler to replace the no-argument `EngineContext` constructor with a constructor that takes a `TaskRunners`?

All users of `EngineContext` must provide a valid `TaskRunners`.  The other fields of `EngineContext` can be default initialized and populated as needed for scenarios like the unit tests.

Reply:
On second thought, I like this. The original intent was to avoid having to define empty TaskRunners everywhere, but it just happens in some tests.

I just added a param to the ctor/made task_runners a const lref/reverted the changes to `TaskRunners`.